x-pack/heartbeat/monitors/browser/source/local.go

```diff
@@ -0,0 +1,148 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package source
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+
+	"github.com/elastic/beats/v7/libbeat/logp"
+
+	"github.com/otiai10/copy"
+)
+
+type LocalSource struct {
+	OrigPath    string `config:"path"`
+	workingPath string
+	BaseSource
+}
+
+var ErrNoPath = fmt.Errorf("local source defined with no path specified")
+
+func ErrInvalidPath(path string) error {
+	return fmt.Errorf("local source has invalid path '%s'", path)
+}
+
+func (l *LocalSource) Validate() error {
+	if l.OrigPath == "" {
+		return ErrNoPath
+	}
+
+	s, err := os.Stat(l.OrigPath)
+	base := ErrInvalidPath(l.OrigPath)
+	if err != nil {
+		return fmt.Errorf("%s: %w", base, err)
+	}
+	if !s.IsDir() {
+		return fmt.Errorf("%s: path points to a non-directory", base)
+	}
+
+	return nil
+}
+
+var offlineEnvVar = "ELASTIC_SYNTHETICS_OFFLINE"
+
+func offline() bool {
+	return os.Getenv(offlineEnvVar) == "true"
+}
+
+func (l *LocalSource) Fetch() (err error) {
+	if l.workingPath != "" {
+		return nil
+	}
+	l.workingPath, err = ioutil.TempDir("/tmp", "elastic-synthetics-")
+	if err != nil {
+		return fmt.Errorf("could not create tmp dir: %w", err)
+	}
+	defer func() {
+		if err != nil {
+			err := l.Close() // cleanup the dir if this function returns an err
+			if err != nil {
+				logp.Warn("could not cleanup dir: %s", err)
+			}
+		}
+	}()
+
+	err = copy.Copy(l.OrigPath, l.workingPath)
+	if err != nil {
+		return fmt.Errorf("could not copy suite: %w", err)
+	}
+
+	dir, err := getAbsoluteSuiteDir(l.workingPath)
+	if err != nil {
+		return err
+	}
+
+	if !offline() {
+		err = setupOnlineDir(dir)
+		return err
+	}
+
+	return nil
+}
+
+// setupOnlineDir is run in environments with internet access and attempts to make sure the node env
+// is setup correctly.
+func setupOnlineDir(dir string) (err error) {
```

Comment:
I don't have tests for this because this requires an internet connection and NPM to be functioning. Our E2E tests in synthetics will ultimately catch issues here. See https://github.com/elastic/beats/issues/23346